src/profile-logic/profile-data.js

```diff
@@ -417,86 +417,97 @@ export function collapsePlatformStackFrames(thread: Thread): Thread {
   });
 }
 
+export function filterThreadToSearchStrings(
+  thread: Thread,
+  searchStrings: string[]
+): Thread {
+  return timeCode('filterThreadToSearchStrings', () => {
+    if (!searchStrings.length) {
+      return thread;
+    }
+
+    return searchStrings.reduce(filterThreadToSearchString, thread);
+  });
+}
+
 export function filterThreadToSearchString(
   thread: Thread,
   searchString: string
 ): Thread {
-  return timeCode('filterThreadToSearchString', () => {
-    if (searchString === '') {
-      return thread;
+  if (!searchString) {
+    return thread;
+  }
+  const lowercaseSearchString = searchString.toLowerCase();
+  const {
+    samples,
+    funcTable,
+    frameTable,
+    stackTable,
+    stringTable,
+    resourceTable,
+  } = thread;
+
+  function computeFuncMatchesFilter(func) {
+    const nameIndex = funcTable.name[func];
+    const nameString = stringTable.getString(nameIndex);
```

Comment:
so, I tried your suggestion about looping through all strings here (with a regexp instead of a loop which is the same thing) but this actually implements a "or" and we want a "and" here.

I've looked at doing a "and" with only one tree traversal but it adds quite a bit of complexity especially with the caching part.

I've looked at the performance impact in a big profile, this doesn't look so bad to me: from my simple test, this goes from ~120ms for 1 term to 140ms for 5 terms. This is of course highly dependent on whether the first term filters a lot of things.

So I'd rather keep this simple approach for now. If we want to make it better in the future I'd rather try to memoize the result of each reduce step, but I don't see it's needed now.

Reply:
So nothing changed from the previous patch: it's only an indentation change + the `reduce` call above

Reply:
Works for me. Thanks for looking into it and profiling it.